plugins/crypto/mbedtls/securitypolicy_mbedtls_common.c

```diff
@@ -289,6 +300,252 @@ mbedtls_decrypt_rsaOaep(mbedtls_pk_context *localPrivateKey,
+
+    /* Get X509 certificate */
+    mbedtls_x509_crt x509Cert;
+    mbedtls_x509_crt_init(&x509Cert);
+    UA_ByteString certificateStr = UA_mbedTLS_CopyDataFormatAware(&securityPolicy->localCertificate);
+    ret = mbedtls_x509_crt_parse(&x509Cert, certificateStr.data, certificateStr.length);
+    UA_ByteString_clear(&certificateStr);
+    if(ret)
+        return UA_STATUSCODE_BADCERTIFICATEINVALID;
+
+    mbedtls_x509write_csr request;
+    mbedtls_x509write_csr_init(&request);
+    /* Set message digest algorithms in CSR context */
+    mbedtls_x509write_csr_set_md_alg(&request, MBEDTLS_MD_SHA256);
+
+    /* Set key usage in CSR context */
+    if(mbedtls_x509write_csr_set_key_usage(&request, MBEDTLS_X509_KU_DIGITAL_SIGNATURE |
+                                                     MBEDTLS_X509_KU_DATA_ENCIPHERMENT |
+                                                     MBEDTLS_X509_KU_NON_REPUDIATION |
+                                                     MBEDTLS_X509_KU_KEY_ENCIPHERMENT) != 0) {
+        retval = UA_STATUSCODE_BADINTERNALERROR;
+        goto cleanup;
+    }
+
+    /* Add entropy */
+    if(entropyContext && nonce && nonce->length > 0) {
+        if(mbedtls_entropy_update_manual(entropyContext,
+                                         (const unsigned char*)(nonce->data),
+                                         nonce->length) != 0) {
+            retval = UA_STATUSCODE_BADINTERNALERROR;
+            goto cleanup;
+        }
+    }
+
+    /* Get subject from argument or read it from certificate */
+    if(subjectName && subjectName->length > 0) {
+        /* subject from argument */
+        subj = (char *)UA_malloc(subjectName->length + 1);
+        if(!subj) {
+            retval = UA_STATUSCODE_BADOUTOFMEMORY;
+            goto cleanup;
+        }
+        memset(subj, 0x00, subjectName->length + 1);
+        strncpy(subj, (char *)subjectName->data, subjectName->length);
+        /* search for / in subject and replace it by comma */
+        char *p = subj;
+        for(size_t i = 0; i < subjectName->length; i++) {
+            if(*p == '/' ) {
+                *p = ',';
+            }
+            ++p;
+        }
+    } else {
+        /* read subject from certificate */
+        mbedtls_x509_name s = x509Cert.subject;
+        subj = (char *)UA_malloc(UA_MAXSUBJECTLENGTH);
+        if(!subj) {
+            retval = UA_STATUSCODE_BADOUTOFMEMORY;
+            goto cleanup;
+        }
+        if(mbedtls_x509_dn_gets(subj, UA_MAXSUBJECTLENGTH, &s) <= 0) {
+            retval = UA_STATUSCODE_BADINTERNALERROR;
+            goto cleanup;
+        }
+    }
+
+    /* Set the subject in CSR context */
+    if(mbedtls_x509write_csr_set_subject_name(&request, subj) != 0) {
+        retval = UA_STATUSCODE_BADINTERNALERROR;
+        goto cleanup;
+    }
+
+    /* Get the subject alternate names from certificate and set them in CSR context*/
+    const mbedtls_x509_sequence *san_list = &x509Cert.subject_alt_names;
```

Comment:
Some compilers complain when you `goto` over a variable declaration.
I would declare the san_list variable at the top / before the first goto.